tests/php/Analyzer/TupleSymbol/ForeachSymbolTest.php

```diff
@@ -0,0 +1,129 @@
+<?php
+
+declare(strict_types=1);
+
+namespace PhelTest\Analyzer\TupleSymbol;
+
+use Generator;
+use Phel\Analyzer;
+use Phel\Analyzer\TupleSymbol\ForeachSymbol;
+use Phel\Ast\ForeachNode;
+use Phel\Exceptions\PhelCodeException;
+use Phel\GlobalEnvironment;
+use Phel\Lang\Symbol;
+use Phel\Lang\Table;
+use Phel\Lang\Tuple;
+use Phel\NodeEnvironment;
+use PHPUnit\Framework\TestCase;
+
+final class ForeachSymbolTest extends TestCase
+{
+    private Analyzer $analyzer;
+
+    public function setUp(): void
+    {
+        $env = new GlobalEnvironment();
+        $env->addDefinition('phel\\core', Symbol::create('first'), new Table());
+        $env->addDefinition('phel\\core', Symbol::create('next'), new Table());
+        $this->analyzer = new Analyzer($env);
+    }
+
+    public function testRequiresAtLeastTwoArg(): void
+    {
+        $this->expectException(PhelCodeException::class);
+        $this->expectExceptionMessage("At least two arguments are required for 'foreach");
+
+        // This is the same as: (foreach)
+        $tuple = Tuple::create(
+            Symbol::create(Symbol::NAME_FOREACH)
+        );
+
+        $this->analyze($tuple);
+    }
+
+    public function testFirstArgMustBeATuple(): void
+    {
+        $this->expectException(PhelCodeException::class);
+        $this->expectExceptionMessage("First argument of 'foreach must be a tuple.");
+
+        // This is the same as: (foreach anything)
+        $mainTuple = Tuple::create(
+            Symbol::create(Symbol::NAME_FOREACH),
+            Symbol::create('anything')
+        );
+
+        $this->analyze($mainTuple);
+    }
+
+    /** @dataProvider providerNumberOfTupleArgs */
+    public function testNumberOfTupleArgs(Tuple $mainTuple, bool $error): void
+    {
+        if ($error) {
+            $this->expectException(PhelCodeException::class);
+            $this->expectExceptionMessage("Tuple of 'foreach must have exactly two or three elements.");
+        } else {
+            self::assertTrue(true); // In order to have an assertion without an error
+        }
+
+        $this->analyze($mainTuple);
+    }
+
+    public function providerNumberOfTupleArgs(): Generator
+    {
+        yield '(foreach (x) x)' => [
+            'mainTuple' => Tuple::create(
+                Symbol::create(Symbol::NAME_FOREACH),
+                Tuple::create(
+                    Symbol::create('x')
+                )
+            ),
+            'error' => true,
+        ];
+
+        yield '(foreach (()))' => [
+            'mainTuple' => Tuple::create(
+                Symbol::create(Symbol::NAME_FOREACH),
+                Tuple::create(
+                    Symbol::create(''),
```

Comment:
Symbol shouldn't be a empty string. From the point of the parser this can never happen. Maybe we need to validate this in the Symbol.

Reply:
Would this be enough/ok?
```php
public function __construct(?string $namespace, string $name)
{
    assertNotEmpty($name, 'Symbol name can not be empty');
```

Reply:
I would throw an `AnalyzerException`.

Reply:
@jenshaase I think I will create another PR for this. Actually, should we only check that the `$name` is not empty, but it can contain any value, right? Or do you have any acceptance-criteria for that in mind?

Reply:
Yes, it can contain any other value.